### ISSUE TITLE
gqrx does not compile against latest gnuradio 3.8 tech preview

### ISSUE THREAD

**Issue by lukaskuzmiak** [csete/gqrx#657]
and I fixed that here: https://github.com/lukaskuzmiak/gqrx/commit/9752cc315da23f2fc99411a7edc360b9145cbf77

It seems you keep master working against gnuradio 3.7 and this will break it, do I create a pull request for master or some other branch (did not find one that would work for that).

Thanks!

**Comment by lukaskuzmiak** [comment]
btw, this has one more issue, due to changes here: https://github.com/gnuradio/gnuradio/commit/6ae02d1d6d04a5f529ab3fea496c589f8d14229d#diff-9e599a674a0b50c90760539b4f037cef

GQRX fails to start with:

```
gr::log :WARN: file_source0 - file is too small
terminate called after throwing an instance of 'std::runtime_error'
  what():  file is too small
Aborted (core dumped)
```

This is because the way they check the file size does not work for `/dev/random`, I verified passing a file w/ size that is dividable by 8 (`sizeof(gr_complex)`) fixes this issue but not really sure what would be a good fix, suggestions welcome.

**Comment by csete** [comment]
This will be a very annoying maintenance problem for many years :(

I could create a separate branch; however, that would require keeping it up to date all the time. So, I suggest that we check the GNU Radio version in cmake, and use #ifdefs in the code to make it build with both version 3.7 and 3.8.

What do you think?

**Comment by lukaskuzmiak** [comment]
Makes sense to me, I'm honestly not sure how many breaking changes is 3.8 going to bring in the APIs and how "spaghetti code" will we need to produce to work around that. But so far it does not look too bad, so I agree #ifdefs sound better than a branch somebody needs to take care of.

**Comment by englishm** [comment]
GNU Radio 3.8.0.0 was released today: 
https://github.com/gnuradio/gnuradio/releases/tag/v3.8.0.0

There is a migration guide for out of tree (OOT) modules here: https://wiki.gnuradio.org/index.php/GNU_Radio_3.8_OOT_Module_Porting_Guide

There's an open bug to update gr-osmosdr here:
https://osmocom.org/issues/3855

And the patches here appear to work (at least with the rc's, I haven't tested again yet):
https://github.com/igorauad/gr-osmosdr/tree/gr3.8

I'm hoping to get gqrx working on FreeBSD again. First we'll need to get the [`comms/gnuradio`](https://www.freshports.org/comms/gnuradio/) and [`comms/gr-osmosdr`](https://www.freshports.org/comms/gr-osmosdr/) ports updated, but then we'll be looking at [`comms/gqrx`](https://www.freshports.org/comms/gqrx).

**Comment by englishm** [comment]
As for the approach of supporting both 3.7 and 3.8... 

The upstream recommendation in that migration guide is to simply create a `maint-3.7` branch, and then move new releases to supporting _only_ 3.8.

> ## [Development Model](https://wiki.gnuradio.org/index.php/GNU_Radio_3.8_OOT_Module_Porting_Guide#Development_Model)
> GNU Radio recently changed the development model. The longstanding next branch was finally merged in to master. From now on, there will only be a maint-X.Y and a master branch. Releasing 3.8 basically means master will be spin off maint-3.8 with the 3.8.0.0 release commit. After that API-breaking changes for 3.9 will end up in master and 3.8 will be maintained in maint-3.8.
> 
> We suggest, you adapt a similar model in your OOT. That means your maint-3.8 branch is compatible with GNU Radio maint-3.8 and your master branch is compatible GNU Radio master. Trying to maintain a single branch that works with all future, present, and past versions of GNU Radio versions is mess and you’d have to differentiate between too many cases (API changes, XML vs YAML GRC bindings, Python 2 vs 3, QT4 vs QT5, log4cpp support or not, CPPUnit or Boost unit tests, etc.).
> 
> Starting your maint-3.7 branch and tagging the current state as the last one that’s compatible with GR 3.7 could be done with:
> 
> ```
> git tag v3.7
> git checkout -b maint-3.7
> ```

Maintaining support for both 3.7 and 3.8 in gqrx would be a really nice feature, but may become a maintenance nightmare. On the other hand, maintaining and releasing from multiple branches has its own costs.

**Comment by englishm** [comment]
It looks like @argilo had good luck getting GQRX to run with the patched version of gr-osmosdr and the changes in this commit:
https://github.com/argilo/gqrx/commit/a02d54cf59ba978d1263b741495a401995ee5df2

https://twitter.com/argilo/status/1156738583010709504/

**Comment by csete** [comment]
Hi @englishm,

I am okay with having separate branches, although if the only changes required in gqrx are some include statements, I think I can easily make it so that the same source code can be compiled with both 3.7 and 3.8.

**Comment by GHPS** [comment]
An update: I just successfully compiled gqrx with the latest versions of
- GNURadio 3.9.0.0 (41348a2b6f7cdb200a88dbcfba936e5a438094c6)
- gr_osmosdr (4d83c6067f059b0c5015c3f59f8117bbd361e877)
and the patched files from PR #696.
- Python 3.6.7 (only, no Python 2 libraries installed)  without PyBOMBS
- Kubuntu 18.04 LTS

GNURadio compiles without any issues. 

gr_osmosdr had to be convinced to cooperate 
(the GNURADIO_RUNTIME_FOUND  function does 
not work, different versions of GrMiscUtils.cmake get
mixed up and the Python sources are still Python2) but
now everything runs smoothly.

**Comment by csete** [comment]
There are several forks of gr-osmosdr that have a gr3.8 branch:

* https://github.com/igorauad/gr-osmosdr/tree/gr3.8
* https://github.com/mvaenskae/gr-osmosdr/tree/gr3.8

**Comment by GHPS** [comment]
> https://github.com/mvaenskae/gr-osmosdr/tree/gr3.8

After a quick test: This fork compiles fine without any of
the issues mentioned above - just out of the box.

**Comment by martymac** [comment]
Hello,

FYI, I've just committed updates to comms/gr-osmosdr and comms/gqrx FreeBSD ports.

(and as already written here: https://osmocom.org/issues/3855, thanks a lot for your hard work, comments and links, it helped a lot!)

Cheers,

Ganael.

**Comment by barjac** [comment]
I just updated our gnuradio package to 3.8 with qt5 and python3 and am now regretting it.
Any idea whether sources ported to match the above are likely in the next 9 months which is when our next distro release of Mageia 8 is due?  We have dropped python2 for this release so unless packages can be ported to 3 then they will have to be dropped. I maintain (and use) gqrx so this would be a sad situation.

**Comment by csete** [comment]
Current git master builds with both GNU Radio 3.7 and 3.8

**Comment by yuvadm** [comment]
@csete thanks for the 2.12 release! :tada:

**Comment by gvanem** [comment]
It seems to have gone from bad to worse. Now all use of:
```c
typedef boost::shared_ptr<decoder> sptr;
```
does not work. One needs to use:
```c
typedef std::shared_ptr<decoder> sptr;
```

Is there a simple fix for that?

**Comment by argilo** [comment]
@gvanem I fixed that in #799.
